### PR TITLE
feature(tx-backend): #536 draft subsequential flow

### PR DESCRIPTION
**Pull request by ds-ext-sceronik**
resolves eclipse-tractusx/traceability-foss#536

### CODE REVIEW

File: tx-backend/src/main/resources/application.yml [hunk 1]
```diff
@@ -47,6 +47,9 @@ edc:
   api-auth-key: ${EDC_API_KEY}
   bpn-provider-url-mappings: { }
   provider-edc-url: ${EDC_PROVIDER_URL}
+  parts-provider-edc-controlplane-url: ${EDC_PARTS_PROVIDER_CONTROLPLANE_URL:https://provider-edc-controlplane.net}
```
Comment (ds-mwesener) [inline tx-backend/src/main/resources/application.yml:50]: Add provider-edc-dataplane-url: ${EDC_PROVIDER_DATAPLANE_URL}

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/application/importpoc/PublishService.java [hunk 2]
```diff
@@ -18,9 +18,13 @@
...
 
+import org.eclipse.tractusx.traceability.assets.domain.base.model.AssetBase;
+
 import java.util.List;
 
 public interface PublishService {
 
     void publishAssets(String policyId, List<String> assetIds);
+
+    void publishAssetsToCx(List<AssetBase> assets);
```
Comment (ds-lcapellino) [inline tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/application/importpoc/PublishService.java:29]:
```suggestion
    void publishAssetsToCoreServices(List<AssetBase> assets);
```

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/base/model/ImportNote.java [hunk 3]
```diff
@@ -24,5 +24,5 @@ public class ImportNote {
     public static final String PERSISTENT_NO_UPDATE = "Asset in sync with digital twin registry. Twin will not be updated.";
     public static final String PERSISTED = "Asset created/updated successfully in persistant state.";
     public static final String IN_SYNCHRONIZATION = "Twin in sync with digital twin registry. Twin will not be updated.";
-
+    public static final String PUBLISHED_TO_CX = "Assets Pubilshed to network";
```
Comment (ds-lcapellino) [inline tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/base/model/ImportNote.java:27]:
```suggestion
    public static final String PUBLISHED_TO_CORE_SERVICES = "Assets Published to network";
```

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/base/model/ImportState.java [hunk 4]
```diff
@@ -20,5 +20,10 @@
 
 
 public enum ImportState {
-    TRANSIENT, PERSISTENT, ERROR, IN_SYNCHRONIZATION, UNSET
+    TRANSIENT,
+    PERSISTENT,
+    ERROR,
+    IN_SYNCHRONIZATION,
+    PUBLISHED_TO_CX,
```
Comment (ds-lcapellino) [inline tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/base/model/ImportState.java:27]:
```suggestion
    PUBLISHED_TO_CORE_SERVICES,
```

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/PolicyServiceImpl.java [hunk 5]
```diff
@@ -36,11 +38,23 @@
...
+        return IrsPolicyResponse.toResponse(getAcceptedPoliciesOrEmptyList());
+    }
 
-        return IrsPolicyResponse.toResponse(allPolicies);
+    @Override
+    public PolicyResponse getPolicyById(String id) {
+        return getAcceptedPoliciesOrEmptyList().stream()
+                .filter(policy -> policy.payload().policy().getPolicyId().equals(id)).findFirst()
+                .map(IrsPolicyResponse::toResponse)
+                .orElseThrow(() -> new PolicyNotFoundException("Policy with id: %s not found.".formatted(id)));
```
[the remark this answers is not in the document]
Comment (ds-ext-sceronik) [reply tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/PolicyServiceImpl.java:52]: The other method is used also in getAllPolicies this one filters by specific assetId and if expected policy is not there then throws exception

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/PublishServiceImpl.java [hunk 6]
```diff
@@ -41,25 +43,38 @@ public class PublishServiceImpl implements PublishService {
...
 
+        //Update assets with policy id
         log.info("Updating status of asPlannedAssets.");
-        updateAssetWithStatusAndPolicy(policyId, assetIds, assetAsPlannedRepository);
+        List<AssetBase> updatedAsPlannedAssets = updateAssetWithStatusAndPolicy(policyId, assetIds, assetAsPlannedRepository);
         log.info("Updating status of asBuiltAssets.");
-        updateAssetWithStatusAndPolicy(policyId, assetIds, assetAsBuiltRepository);
+        List<AssetBase> updatedAsBuiltAssets = updateAssetWithStatusAndPolicy(policyId, assetIds, assetAsBuiltRepository);
+
+        //Publish to CS network
```
[the remark this answers is not in the document]
Comment (ds-ext-sceronik) [reply tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/PublishServiceImpl.java:59]: removed comment as method name is descriptive enough :)

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/common/config/ApplicationConfig.java [hunk 7]
```diff
@@ -146,4 +156,24 @@ public EdcPolicyDefinitionService edcPolicyDefinitionService(EdcConfiguration ed
     public EdcContractDefinitionService edcContractDefinitionService(EdcConfiguration edcConfiguration, RestTemplate edcNotificationAssetRestTemplate) {
         return new EdcContractDefinitionService(edcConfiguration, edcNotificationAssetRestTemplate);
     }
+
+    @Bean
+    public EdcAssetService edcDtrAssetService(EdcConfiguration edcConfiguration, EdcTransformer edcTransformer, RestTemplate edcDtrAssetRestTemplate) {
```
[the remark this answers is not in the document]
Comment (ds-ext-sceronik) [reply tx-backend/src/main/java/org/eclipse/tractusx/traceability/common/config/ApplicationConfig.java:161]: Yep agree. have not considered it in previous implementation. good catch ! for now added EdcConfiguration class that holds all EdcRelated beans that are created from irs-edc-lib :)

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/AsyncPublishService.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    private final AssetAsBuiltRepository assetAsBuiltRepository;
+    private final EdcAssetCreationService edcAssetCreationService;
+    private final DtrService dtrService;
+    private final DecentralRegistryServiceImpl decentralRegistryService;
+
+    @Async(value = AssetsAsyncConfig.PUBLISH_ASSETS_EXECUTOR)
+    public void publishAssetsToCoreServices(List<AssetBase> assets, boolean triggerSynchronizeAssets) {
+        Map<String, List<AssetBase>> assetsByPolicyId = assets.stream().collect(Collectors.groupingBy(AssetBase::getPolicyId));
+
+        List<String> createdShellsAssetIds = new ArrayList<>(List.of());
```
Comment (ds-mwesener) [inline tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/AsyncPublishService.java:55]: List.of() can be removed.

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/AsyncPublishService.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        assetsByPolicyId.forEach((policyId, assetsForPolicy) -> {
+            String submodelServerAssetId = edcAssetCreationService.createDtrAndSubmodelAssets(policyId);
+            assetsForPolicy.forEach(assetBase -> {
+                try {
+                    String assetId = dtrService.createShellInDtr(assetBase, submodelServerAssetId);
+                    createdShellsAssetIds.add(assetId);
+                } catch (CreateDtrShellException e) {
+                    log.error("Failed to create shell in dtr for asset with id %s".formatted(assetBase.getId()), e);
```
Comment (ds-mwesener) [inline tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/AsyncPublishService.java:63]: Shouldnt we have some kind of an error somewhere? I only see an log error message.

File: tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/PolicyServiceImpl.java [hunk 10]
```diff
@@ -36,11 +38,23 @@
...
+    }
+
+    @NotNull
+    private List<IrsPolicyResponse> getAcceptedPoliciesOrEmptyList() {
+        return Optional.ofNullable(irsRepository.getPolicies())
```
Comment (ds-mwesener) [inline tx-backend/src/main/java/org/eclipse/tractusx/traceability/assets/domain/importpoc/service/PolicyServiceImpl.java:57]: emptyIfNull(irsRepository.getPolicies());

**Review by ds-mwesener — CHANGES_REQUESTED** [review]
Please fix.

**Pull request merged**